File: python/starfield.py
import random, time, os

DELAY = 0.02

def main():
    change_amount = 0.5
    density = 0.0
    while True:
        width = os.get_terminal_size()[0] - 1
        if density < 0 or density > 100:
            change_amount *= -1
        density = density + change_amount

        line = ''
        for i in range(width):
            if random.randint(0, 100) < density:
                line = line + '*'
            else:
                line = line + ' '

        print(line); time.sleep(DELAY)

try:
    main()
except KeyboardInterrupt:
    print('Starfield, by Al Sweigart al@inventwithpython.com')

File: typescript/starfield.ts
// To get ts types, run: npm i --save-dev @types/readline

namespace Starfield {
    const DELAY: number = 20;
    const STAR_CHAR: string = '*';
    const EMPTY_CHAR: string = ' ';

    let change_amount: number = 0.005;
    let density: number = 0.0;
    let width: number = 120
    let line: string = '';

    async function sleep(milliseconds: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, milliseconds));
    }

    process.on('SIGINT', function() {
        console.log("\nStarfield by Al Sweigart al@inventwithpython.com");
        process.exit();
    });

    async function main(): Promise<void> {
        while (true) {
            if (process.stdout.isTTY) {
                width = process.stdout.columns - 1; // -1 is because Windows puts a newline when printing to the last column
            }

            if (density < 0 || density > 1) {
                change_amount *= -1;
            }
            density += change_amount;

            line = '';
            for (let i = 0; i < width; i++) {
                if (Math.random() < density) {
                    line += STAR_CHAR;
                } else {
                    line += EMPTY_CHAR;
                }
            }
            console.log(line);
            await sleep(DELAY);
        }
    }

    main();
}